[Curl] Discard response headers when an error occurs

kshukuwa opened the pull request:
#### c9f5e13a69fa0a4098f85b6bf73ababb5e56aa07
<pre>
[Curl] Discard response headers when an error occurs
<a href="https://bugs.webkit.org/show_bug.cgi?id=247716">https://bugs.webkit.org/show_bug.cgi?id=247716</a>

Reviewed by Fujii Hironori.

If a cURL error occurs before the response body is received,
the previously received response headers are not notified to WebCore
to avoid malfunction.

* Source/WebCore/platform/network/curl/CurlRequest.cpp:
(WebCore::CurlRequest::didReceiveHeader):
(WebCore::CurlRequest::didCompleteTransfer):
(WebCore::CurlRequest::shouldAbortConnectRequest): Deleted.
* Source/WebCore/platform/network/curl/CurlRequest.h:

Canonical link: <a href="https://commits.webkit.org/256639@main">https://commits.webkit.org/256639@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d5078b40b826d0c2d00ca36736155f105245a9be

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96402 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5656 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29468 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105945 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166293 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100385 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5827 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34402 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88781 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/102671 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102073 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83004 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31320 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88071 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74211 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/40132 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19563 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37806 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20951 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4607 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/3186 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43520 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/723 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40220 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebCore/platform/network/curl/CurlRequest.h:
```diff
@@ -150,6 +150,8 @@ class CurlRequest : public ThreadSafeRefCounted<CurlRequest>, public CurlRequest
     void updateHandlePauseState(bool);
     bool isHandlePaused() const;
 
+    bool shouldAbortConnectRequest(const URL&, long, long);
```
Please add argument names `statusCode` and `httpConnectCode` here.

fujii replied:
But, I don't think a new member function should be added for this.
A local variable or local lambda function seem enough.

bool shouldAbortConnectRequest = ...;
if (shouldAbortConnectRequest)

auto shouldAbortConnectRequest = [] { return ... };
if (shouldAbortConnectRequest())

WDYT?

kshukuwa replied:
> bool shouldAbortConnectRequest = ...;
> if (shouldAbortConnectRequest)

Yeah. This is simpler and better.

fujii commented on Source/WebCore/platform/network/curl/CurlRequest.cpp:
```diff
@@ -748,6 +751,14 @@ bool CurlRequest::isHandlePaused() const
 
+bool CurlRequest::shouldAbortConnectRequest(const URL& url, long statusCode, long httpConnectCode)
+{
+    if (url.protocolIs("https"_s) && !statusCode && httpConnectCode && httpConnectCode != 200 && httpConnectCode != 407)
```
How does `curl` command handle the case? Does `curl` command just ignore the connect error? Is this a curl bug?

fujii replied:
Do you think "https" check is needed here? I guess httpConnectCode is zero in case of http. Right? If httpConnectCode is zero, this check fails as expected.

kshukuwa replied:
> How does curl command handle the case? Does curl command just ignore the connect error? Is this a curl bug?

Good point! The curl handle those cases correctly. WebKit's curl port seems to be misimplemented.

```
> curl --version
curl 7.85.0 (x86_64-w64-mingw32) libcurl/7.85.0 OpenSSL/1.1.1q (Schannel) zlib/1.2.12 brotli/1.0.9 zstd/1.5.2 libidn2/2.3.3 libssh2/1.10.0 nghttp2/1.48.0
Release-Date: 2022-08-31
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz MultiSSL NTLM SPNEGO SSL SSPI threadsafe TLS-SRP zstd

> curl https://webkit.org -x http://127.0.0.1:8888
curl: (56) Received HTTP code 302 from proxy after CONNECT
```

https://github.com/curl/curl/blob/curl-7_85_0/lib/http_proxy.c#L670-L694
```
  if(data->info.httpproxycode/100 != 2) {
    ...
    failf(data, "Received HTTP code %d from proxy after CONNECT",
          data->req.httpcode);
    return CURLE_RECV_ERROR;
  }
```

If the result of `CurlRequest::didCompleteTransfer` is not `CURLE_OK`, it seems that `invokeDidReceiveResponse` should not be executed.

https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/network/curl/CurlRequest.cpp#L469-L476
```
void CurlRequest::didCompleteTransfer(CURLcode result)
{
   ...

    if (needToInvokeDidReceiveResponse()) {
        // Processing of didReceiveResponse() has not been completed. (For example, HEAD method)
        // When completeDidReceiveResponse() is called, didCompleteTransfer() will be called again.

        m_finishedResultCode = result;
        invokeDidReceiveResponse(m_response, Action::FinishTransfer);  <<== HERE
        return;
    }
    ...
}
```

fujii reviewed (commented):
Sorry, I don't understand. Why did a problem happen if `invokeDidReceiveResponse` was called in this case?

kshukuwa committed:
[Curl] Discard response headers when an error occurs
https://bugs.webkit.org/show_bug.cgi?id=247716

Reviewed by Fujii Hironori.

If a cURL error occurs before the response body is received,
the previously received response headers are not notified to WebCore
to avoid malfunction.

* Source/WebCore/platform/network/curl/CurlRequest.cpp:
(WebCore::CurlRequest::didReceiveHeader):
(WebCore::CurlRequest::didCompleteTransfer):
(WebCore::CurlRequest::shouldAbortConnectRequest): Deleted.
* Source/WebCore/platform/network/curl/CurlRequest.h:

Canonical link: https://commits.webkit.org/256639@main
